File: mobile/test/logic/receipt_parser_test.dart
import 'package:cash_compass/logic/receipt_parser.dart';
import 'package:cash_compass/logic/subscriptions.dart';
import 'package:cash_compass/models/transaction.dart';
import 'package:flutter_test/flutter_test.dart';

/// Fixture OCR text rather than real images.
///
/// This isolates the extraction rules from ML Kit entirely: no camera, no
/// emulator, no native dependency, and a failure points at the parser rather
/// than at the recognizer.
void main() {
  ParsedReceipt parse(String text) =>
      parseReceiptLines(text.trim().split('\n'));

  group('amount extraction', () {
    test('prefers a labelled total', () {
      final r = parse('''
BLUE BOTTLE COFFEE
123 Market Street
Latte            4.50
Croissant        3.25
Subtotal         7.75
Tax              0.62
Total            8.37
''');
      expect(r.amount, 8.37);
      expect(r.amountConfidence, FieldConfidence.high);
    });

    test('ignores subtotal and tax lines', () {
      final r = parse('''
Shop
Subtotal        99.99
Tax              5.00
Total           12.34
''');
      expect(r.amount, 12.34, reason: 'subtotal must not win over total');
    });

    test('takes the last total when a tip line follows', () {
      // Receipts print the pre-tip total first, so reading bottom-up matters.
      final r = parse('''
The Grill House
Total            40.00
Tip               6.00
Total after tip  46.00
''');
      expect(r.amount, 46.00);
    });

    test('handles currency symbols and thousands separators', () {
      final r = parse('''
Electronics World
Grand Total    \$1,299.99
''');
      expect(r.amount, 1299.99);
      expect(r.amountConfidence, FieldConfidence.high);
    });

    test('reads the value from the line below the label', () {
      final r = parse('''
Corner Store
TOTAL
15.75
''');
      expect(r.amount, 15.75);
      expect(r.amountConfidence, FieldConfidence.medium);
    });

    test('falls back to the largest number, marked low confidence', () {
      final r = parse('''
Mystery Shop
item one      3.00
item two     11.50
item three    7.25
''');
      expect(r.amount, 11.50);
      expect(
        r.amountConfidence,
        FieldConfidence.low,
        reason: 'a guess must not look as certain as a labelled total',
      );
    });
  });

  group('merchant extraction', () {
    test('takes the top line and title-cases it', () {
      final r = parse('''
BLUE BOTTLE COFFEE
123 Market Street
Total 8.37
''');
      expect(r.merchant, 'Blue Bottle Coffee');
      expect(r.merchantConfidence, FieldConfidence.high);
    });

    test('skips phone numbers and addresses', () {
      final r = parse('''
+44 20 7946 0958
14 Baker Street
Tesco Express
Total 22.10
''');
      expect(r.merchant, 'Tesco Express');
      expect(r.merchantConfidence, FieldConfidence.medium);
    });

    test('skips lines that are mostly digits', () {
      final r = parse('''
20260821 0931 4471
Pharmacy Plus
Total 9.99
''');
      expect(r.merchant, 'Pharmacy Plus');
    });
  });

  group('category guessing', () {
    test('maps a merchant keyword to a category', () {
      expect(parse('Joe\'s Pizza Kitchen\nTotal 20.00').category, 'Food');
      expect(parse('Uber Trip\nTotal 14.20').category, 'Transport');
      expect(parse('CVS Pharmacy\nTotal 8.10').category, 'Health');
      expect(parse('Tesco Supermarket\nTotal 45.00').category, 'Groceries');
    });

    test('defaults to Other rather than guessing', () {
      final r = parse('Zylnor Ltd\nTotal 5.00');
      expect(r.category, defaultReceiptCategory);
      expect(r.categoryConfidence, FieldConfidence.none);
    });

    test('every mapped category is a real expense category', () {
      // Guards against the keyword map drifting away from the app's list.
      for (final category in receiptCategoryTargets) {
        expect(
          expenseCategories,
          contains(category),
          reason: '"$category" is not in expenseCategories',
        );
      }
    });

    test('prefers the longer keyword on overlap', () {
      // "gas station" -> Transport must beat a bare substring match.
      expect(parse('Shell Gas Station\nTotal 60.00').category, 'Transport');
    });
  });

  group('unusable input falls through to manual entry', () {
    test('empty text yields an empty result', () {
      final r = parse('\n   \n');
      expect(r.isEmpty, isTrue);
      expect(r.amount, isNull);
      expect(r.merchant, isNull);
    });

    test('garbled OCR yields no amount', () {
      final r = parse('~~~ \$\$\$ ### ??? ***');
      expect(r.amount, isNull);
      expect(r.amountConfidence, FieldConfidence.none);
    });

    test('non-Latin script is treated as unreadable, not a crash', () {
      // ML Kit's Latin recognizer returns little or nothing for these; the
      // parser must degrade rather than throw.
      final r = parse('スーパーマーケット\n合計 1200');
      expect(r.merchant, isNull, reason: 'no Latin letters to work with');
      expect(() => r.isEmpty, returnsNormally);
    });
  });

  group('bounds', () {
    test('raw text is capped for a very long receipt', () {
      final lines = [
        'Mega Market',
        for (var i = 0; i < 2000; i++) 'Item $i    ${i % 90 + 1}.99',
        'Total 42.00',
      ];
      final r = parseReceiptLines(lines);
      expect(r.rawText.length, lessThanOrEqualTo(4100));
      expect(r.amount, isNotNull, reason: 'still parses despite the cap');
    });
  });

  group('candidate subscription check', () {
    FinanceTransaction charge(String date, double amount) => FinanceTransaction(
          id: 'h-$date',
          name: 'Netflix',
          amount: amount,
          type: TransactionType.expense,
          category: 'Entertainment',
          date: date,
        );

    test('flags a merchant already charging monthly', () {
      final flagged = wouldBeSubscription(
        history: [charge('2026-06-01', 15.99), charge('2026-07-01', 15.99)],
        merchant: 'Netflix',
        amount: 15.99,
        date: '2026-08-01',
      );
      expect(flagged, isTrue);
    });

    test('does not flag a one-off merchant', () {
      final flagged = wouldBeSubscription(
        history: [charge('2026-06-01', 15.99)],
        merchant: 'Corner Bakery',
        amount: 4.50,
        date: '2026-08-01',
      );
      expect(flagged, isFalse);
    });

    test('does not flag when the amount is wildly different', () {
      final flagged = wouldBeSubscription(
        history: [charge('2026-06-01', 15.99), charge('2026-07-01', 15.99)],
        merchant: 'Netflix',
        amount: 99.00,
        date: '2026-08-01',
      );
      expect(
        flagged,
        isFalse,
        reason: 'the ±15% amount rule must still apply to the candidate',
      );
    });
  });
}

File: frontend/src/test/receiptParser.test.ts
import { describe, expect, it } from "vitest";

import {
  DEFAULT_RECEIPT_CATEGORY,
  isReceiptEmpty,
  parseReceiptLines,
  receiptCategoryTargets,
} from "@/lib/receiptParser";

/**
 * The same fixtures as `mobile/test/logic/receipt_parser_test.dart`.
 *
 * Kept deliberately identical, case for case. The two parsers are separate
 * implementations of one set of rules, and the only thing stopping them
 * drifting apart is that they are held to the same examples. If you change a
 * case here, change it there in the same PR.
 */

const parse = (text: string) => parseReceiptLines(text.split("\n"));

/** Categories the app offers for an expense, mirroring `expenseCategories`
 * in QuickActions.tsx. */
const expenseCategories = [
  "Housing", "Groceries", "Transport", "Entertainment", "Food",
  "Utilities", "Shopping", "Health", "Travel", "Education", "Other",
];

describe("amount extraction", () => {
  it("prefers a labelled total", () => {
    const r = parse(`
BLUE BOTTLE COFFEE
123 Market Street
Latte            4.50
Croissant        3.25
Subtotal         7.75
Tax              0.62
Total            8.37
`);
    expect(r.amount).toBe(8.37);
    expect(r.amountConfidence).toBe("high");
  });

  it("ignores subtotal and tax lines", () => {
    const r = parse(`
Shop
Subtotal        99.99
Tax              5.00
Total           12.34
`);
    expect(r.amount, "subtotal must not win over total").toBe(12.34);
  });

  it("takes the last total when a tip line follows", () => {
    // Receipts print the pre-tip total first, so reading bottom-up matters.
    const r = parse(`
The Grill House
Total            40.00
Tip               6.00
Total after tip  46.00
`);
    expect(r.amount).toBe(46.0);
  });

  it("handles currency symbols and thousands separators", () => {
    const r = parse(`
Electronics World
Grand Total    $1,299.99
`);
    expect(r.amount).toBe(1299.99);
    expect(r.amountConfidence).toBe("high");
  });

  it("reads the value from the line below the label", () => {
    const r = parse(`
Corner Store
TOTAL
15.75
`);
    expect(r.amount).toBe(15.75);
    expect(r.amountConfidence).toBe("medium");
  });

  it("falls back to the largest number, marked low confidence", () => {
    const r = parse(`
Mystery Shop
item one      3.00
item two     11.50
item three    7.25
`);
    expect(r.amount).toBe(11.5);
    expect(
      r.amountConfidence,
      "a guess must not look as certain as a labelled total",
    ).toBe("low");
  });
});

describe("merchant extraction", () => {
  it("takes the top line and title-cases it", () => {
    const r = parse(`
BLUE BOTTLE COFFEE
123 Market Street
Total 8.37
`);
    expect(r.merchant).toBe("Blue Bottle Coffee");
    expect(r.merchantConfidence).toBe("high");
  });

  it("skips phone numbers and addresses", () => {
    const r = parse(`
+44 20 7946 0958
14 Baker Street
Tesco Express
Total 22.10
`);
    expect(r.merchant).toBe("Tesco Express");
    expect(r.merchantConfidence).toBe("medium");
  });

  it("skips lines that are mostly digits", () => {
    const r = parse(`
20260821 0931 4471
Pharmacy Plus
Total 9.99
`);
    expect(r.merchant).toBe("Pharmacy Plus");
  });
});

describe("category guessing", () => {
  it("maps a merchant keyword to a category", () => {
    expect(parse("Joe's Pizza Kitchen\nTotal 20.00").category).toBe("Food");
  });

  it("defaults to Other rather than guessing", () => {
    expect(parse("Zzyzx Holdings\nTotal 5.00").category).toBe(
      DEFAULT_RECEIPT_CATEGORY,
    );
  });

  it("every mapped category is a real expense category", () => {
    for (const category of receiptCategoryTargets()) {
      expect(
        expenseCategories,
        `${category} is not a category the app offers`,
      ).toContain(category);
    }
  });

  it("prefers the longer keyword on overlap", () => {
    expect(parse("Shell Gas Station\nTotal 40.00").category).toBe("Transport");
  });
});

describe("unusable input falls through to manual entry", () => {
  it("empty text yields an empty result", () => {
    const r = parse("");
    expect(isReceiptEmpty(r)).toBe(true);
    expect(r.amount).toBeNull();
  });

  it("garbled OCR yields no amount", () => {
    expect(parse("~~~\n???\n***").amount).toBeNull();
  });

  it("non-Latin script is treated as unreadable, not a crash", () => {
    const r = parse("日本語のレシート\n合計");
    expect(r.amount).toBeNull();
  });
});

describe("bounds", () => {
  it("raw text is capped for a very long receipt", () => {
    const lines = Array.from({ length: 500 }, (_, i) => `item ${i} 1.00`);
    const r = parseReceiptLines(lines);
    expect(
      r.rawText.length,
      "an unbounded receipt would balloon the stored transaction",
    ).toBeLessThanOrEqual(4001);
  });
});
